### PR TITLE
[Feature] Add `temperature` to JSON generator

### PR DESCRIPTION
**Pull request by riedgar-ms**
Use `with_temperature()` to support users setting a temperature when generating JSON. Augment `test_json.py` to ensure that the mocked model got called with the specified temperature.

### CODE REVIEW

**Comment by riedgar-ms** [comment]
Tag @hudson-ai for feedback too

File: tests/library/test_json.py [hunk 1]
```diff
@@ -19,11 +21,34 @@ def _generate_and_check(target_obj: Any, schema_obj):
...
+            name=CAPTURE_KEY, schema=schema_obj, temperature=desired_temperature
+        )
+    else:
+        lm += gen_json(name=CAPTURE_KEY, schema=schema_obj)
 
     # Make sure the round trip works
     assert json.loads(lm[CAPTURE_KEY]) == target_obj
 
+    # Check on some temperatures
+    if desired_temperature is not None:
```
[the remark this answers is not in the document]
Comment (Harsha-Nori) [reply tests/library/test_json.py:35]: I think that's an artifact/limitation of the `Mock` class

**Comment by Harsha-Nori** [comment]
Looks reasonable to me

**Comment by Harsha-Nori** [comment]
At some point, we need a better story for all sampling params (instead of piping them through every function like `gen`, `json`, etc.) but we can cross that bridge in the future. Scott had some initial sketches on API design for that here: https://github.com/guidance-ai/guidance/issues/600#issuecomment-1912861646 @slundberg

**Comment by hudson-ai** [comment]
> At some point, we need a better story for all sampling params (instead of piping them through every function like `gen`, `json`, etc.) but we can cross that bridge in the future. Scott had some initial sketches on API design for that here: [#600 (comment)](https://github.com/guidance-ai/guidance/issues/600#issuecomment-1912861646) @slundberg

@Harsha-Nori great point. It would be nice to come to a consensus on the overall API -- would you mind opening an issue to track that as a broader topic?

@slundberg I have some questions/thoughts about the specific idea around using context managers for this (sorry if this discussion belongs in #600 or the hypothetical issue I just asked to be opened):

I'm not yet super familiar with this corner of how the language is implemented, _but_ it's my understanding that guidance context blocks are implemented by gluing an object to the base `Model` class at `__enter__` time and popping that object off at `__exit__` time (`block` and `role` are the only context managers in the language at the moment?). AFAIK, functions decorated with `stateless=True` are "compiled" by passing a null grammar to the function being decorated, meaning that we don't have access to an actual `Model` object when the expressions within the context block are being evaluated.

All of this is to say that I think that the current approach to context blocks can't actually change generation behavior when they are used from inside stateless grammars, only when they are used in a stateful setting. Is that right?

If this is the case, it seems that this inconsistency between stateful and stateless functions is rather opaque and unintuitive from a user standpoint, and it may require a significant rewrite of the approach to context managers to remedy this.

**Comment by riedgar-ms** [comment]
While improved handling of the model params is certainly desirable, I intend to merge this unless there are immediate objections

**Pull request merged**